### ISSUE TITLE
Questiones about fsdp + pure bf16

### ISSUE THREAD

**Issue by kongjiellx** [meta-llama/llama-recipes#148]
I want to continue pre-training the LLMa70b model in order to add Chinese tokens and train it on Chinese data. I'm considering using FSDP along with pure bf16. However, I have two questions:

1. According to [this GitHub discussion](https://github.com/pytorch/torchdistx/pull/52#discussion_r1087107777), should I set use_kahan_summation=True?
2. Should I use the keep_low_precision_grads option in the FSDP MixedPrecision config?

**Comment by HamidShojanazeri** [comment]
@kongjiellx in case you want to do pure bf16, yes you can set the use_kahan_summation=True [here](https://github.com/facebookresearch/llama-recipes/blob/main/policies/anyprecision_optimizer.py#L24) along with keeping the momentum and variance and the buffer in bf16 [here](https://github.com/facebookresearch/llama-recipes/blob/main/policies/anyprecision_optimizer.py#L25-Lb27) as well. 

Make sure to pass `anyprecision` [here](https://github.com/facebookresearch/llama-recipes/blob/main/configs/fsdp.py#L17) as the optimizer as well.

**Comment by kongjiellx** [comment]
@HamidShojanazeri Is the keep_low_precision_grads option necessary? My understanding is that when keep_low_precision_grads=False (which is the default setting), gradients are cast to float32 before being passed to the optimizer. Is this the expected behavior? This question is related to [GitHub Pull Request #85223](https://github.com/pytorch/pytorch/pull/85223).

**Comment by siddk** [comment]
Follow-up question for @HamidShojanazeri and others maintaining this codebase; can you help explain the difference between using "traditional BF16 mixed precision" and "pure BF16" with/without the AnyPrecisionOptimizer?

Specifically, let's say I want to run full-finetuning for a LLaMa-7B model. There are a couple ways I can instantiate various components:

1. Traditional Mixed Precision (BF16):
```python
model = LlamaModelForCausalLM.from_pretrained(...)

# Wrap in FSDP w/ Mixed Precision Policy `bfSixteen_mixed`
bfSixteen_mixed = MixedPrecision(param_dtype=torch.float32, reduce_dtype=torch.bfloat16, buffer_dtype=torch.bfloat16) 
model = FSDP(model, mixed_precision= bfSixteen_mixed, ...)

# Optimizer 
optimizer = AdamW(model.parameters(), ...)

# Forward Pass (uses torch.autocast)
with torch.autocast(dtype=torch.bfloat16):
      loss = model(batch)

loss.backward()
optimizer.step()
optimizer.zero_grad()
```

2. Pure BF16:

```python
# After the `model = ` line
model.to(dtype=torch.bfloat16)

# Wrap in FSDP
bfSixteenPure = MixedPrecision(param_dtype=torch.bfloat16, reduce_dtype=torch.bfloat16, buffer_dtype=torch.bfloat16)
model = FSDP(model, mixed_precision= bfSixteenPure)

# Optimizer
optimizer = ???
```

From the above thread, it seems like for the pure BF16 case, I should be using the AnyPrecisionOptimizer with `use_kahan_summation=True`; however, I've been using a traditional AdamW optimizer instead, with stable finetuning -- I did this under the assumption that AdamW always keeps momentum/variances in full precision (and applies update in full precision).

Is that assumption wrong? For stability and memory efficiency, which of the above schemes is the "correct" way to finetune?

**Comment by kongjiellx** [comment]
@siddk 
From what I understand, the main difference between the first method you mentioned and the second one is the datatype of the optimizer parameters — Traditional Mixed Precision corresponds to fp32, while pure bf16 corresponds to bf16. The "keep_low_precision_grads" I mentioned earlier prevents upcasting during optimizer calculations. This is primarily designed to save GPU memory, but due to the loss of precision from bf16 compared to fp32, it may affect the results. There is some discussion about the effects of AnyPrecisionAdamW [here](https://github.com/pytorch/torchdistx/pull/52).

However, I haven't fully grasped the impact of the [MixedPrecision](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.MixedPrecision) settings on training outcomes from the documentation. I hope someone can help clarify this. Thank you.

**Comment by kongjiellx** [comment]
Additionally, in my training with the 70b llama model, I observed that the loss reduction was largely comparable between pure bf16 and traditional bf16 mixed precision. (I only observed for a few hundred steps with a batch size of 64.)
